### PR TITLE
Fix library

### PR DESCRIPTION
**Pull request by LocutusOfBorg**
Not yet ready :)

The solution might be:
- stop using the fake calls to printf for debug and use the libettercap version
  DONE: rename the gbls to some other thing, 
- use also the gbls from libettercap (the one you need BTW)
- create a "libettercap_init" with the proper inizialization code
- use "extern" variables to pass something to libettercap when we initialize (e.g. the pointers to something we use in the main)
- use libettercap code with no problems at all
- remove duplicated code (e.g. el_target.c) I tried to do it once if you remember, but I was stopped by the above points.

### CODE REVIEW

**Comment by koeppea** [comment]
Wow this is huge. Looking very forward for this one....

**Comment by koeppea** [comment]
Isn't better to rebase onto master rather than mergin master into this branch: Avoiding merge commits which raise themselfes often issues in rebasing.

**Comment by LocutusOfBorg** [comment]
I guess I did it :)

**Comment by LocutusOfBorg** [comment]
@koeppea I did  a lot of work, but I'm stuck in merging ```target_env``` struct from ec and etterlog
```
ettercap/utils/etterlog/el_parser.c: In function ‘parse_options’:
ettercap/utils/etterlog/el_parser.c:197:32: error: ‘struct target_env’ has no member named ‘proto’
                   EL_GBL_TARGET->proto = strdup(optarg);
                                ^
ettercap/utils/etterlog/el_target.c: In function ‘is_target_pck’:
ettercap/utils/etterlog/el_target.c:282:23: error: ‘struct target_env’ has no member named ‘proto’
     if (!EL_GBL_TARGET->proto || !strcmp(EL_GBL_TARGET->proto, "") || !strcasecmp(EL_GBL_TARGET->proto, "all"))  
                       ^
ettercap/utils/etterlog/el_target.c:282:55: error: ‘struct target_env’ has no member named ‘proto’
     if (!EL_GBL_TARGET->proto || !strcmp(EL_GBL_TARGET->proto, "") || !strcasecmp(EL_GBL_TARGET->proto, "all"))  
                                                       ^
ettercap/utils/etterlog/el_target.c:282:96: error: ‘struct target_env’ has no member named ‘proto’
     if (!EL_GBL_TARGET->proto || !strcmp(EL_GBL_TARGET->proto, "") || !strcasecmp(EL_GBL_TARGET->proto, "all"))  
                                                                                                ^
ettercap/utils/etterlog/el_target.c:285:22: error: ‘struct target_env’ has no member named ‘proto’
     if (EL_GBL_TARGET->proto && !strcasecmp(EL_GBL_TARGET->proto, "tcp") 
                      ^
ettercap/utils/etterlog/el_target.c:285:58: error: ‘struct target_env’ has no member named ‘proto’
     if (EL_GBL_TARGET->proto && !strcasecmp(EL_GBL_TARGET->proto, "tcp") 
                                                          ^
ettercap/utils/etterlog/el_target.c:289:22: error: ‘struct target_env’ has no member named ‘proto’
     if (EL_GBL_TARGET->proto && !strcasecmp(EL_GBL_TARGET->proto, "udp") 
                      ^
ettercap/utils/etterlog/el_target.c:289:58: error: ‘struct target_env’ has no member named ‘proto’
     if (EL_GBL_TARGET->proto && !strcasecmp(EL_GBL_TARGET->proto, "udp") 
                                                          ^
ettercap/utils/etterlog/el_target.c: In function ‘is_target_info’:
ettercap/utils/etterlog/el_target.c:352:22: error: ‘struct target_env’ has no member named ‘proto’
    if (!EL_GBL_TARGET->proto || !strcmp(EL_GBL_TARGET->proto, "") || !strcasecmp(EL_GBL_TARGET->proto, "all"))  
                      ^
ettercap/utils/etterlog/el_target.c:352:54: error: ‘struct target_env’ has no member named ‘proto’
    if (!EL_GBL_TARGET->proto || !strcmp(EL_GBL_TARGET->proto, "") || !strcasecmp(EL_GBL_TARGET->proto, "all"))  
                                                      ^
ettercap/utils/etterlog/el_target.c:352:95: error: ‘struct target_env’ has no member named ‘proto’
    if (!EL_GBL_TARGET->proto || !strcmp(EL_GBL_TARGET->proto, "") || !strcasecmp(EL_GBL_TARGET->proto, "all"))  
                                                                                               ^
ettercap/utils/etterlog/el_target.c:361:27: error: ‘struct target_env’ has no member named ‘proto’
          if (EL_GBL_TARGET->proto && !strcasecmp(EL_GBL_TARGET->proto, "tcp") 
                           ^
ettercap/utils/etterlog/el_target.c:361:63: error: ‘struct target_env’ has no member named ‘proto’
          if (EL_GBL_TARGET->proto && !strcasecmp(EL_GBL_TARGET->proto, "tcp") 
                                                               ^
ettercap/utils/etterlog/el_target.c:365:27: error: ‘struct target_env’ has no member named ‘proto’
          if (EL_GBL_TARGET->proto && !strcasecmp(EL_GBL_TARGET->proto, "udp") 
                           ^
ettercap/utils/etterlog/el_target.c:365:63: error: ‘struct target_env’ has no member named ‘proto’
          if (EL_GBL_TARGET->proto && !strcasecmp(EL_GBL_TARGET->proto, "udp") 
                                                               ^
utils/CMakeFiles/etterlog.dir/build.make:215: recipe for target 'utils/CMakeFiles/etterlog.dir/etterlog/el_parser.c.o' failed
```

of course I can add the field, but I'm wondering if we can remove some etterlog code and use the ettercap version

**Comment by koeppea** [comment]
Hi Franky,
nice to hear from you. Happy new 2016.
Honestly I don't know of anybody who ever used *etterlog*.
So I support your approach.

**Comment by LocutusOfBorg** [comment]
The problem is that "proto" is used to understand if we are in tcp or udp, and the target_env from ettercap doesn't have this field, so in order to use it, I need to make etterlog work in a ettercap-similar way. I see an EC_GBL_OPTIONS->proto field, maybe we can use that?

**Comment by koeppea** [comment]
Yes if we could use this, it defenitely fits it's use-case.

**Comment by eaescob** [comment]
Any updates?

**Comment by LocutusOfBorg** [comment]
this is a very little part of my work, it is really an huge stuff to do

**Comment by koeppea** [comment]
Wow this is huge.
Let me test this pull.

**Comment by koeppea** [comment]
Just tested.
Build goes fine.
But linking fails. Seems only _interfaces_ and _libpcap_ has issues:
```
Linking C executable ettercap
CMakeFiles/ettercap.dir/ec_parser.c.o: In function `parse_options':
/home/koeppea/dev/ettercap/src/ec_parser.c:256: undefined reference to `select_text_interface'
/home/koeppea/dev/ettercap/src/ec_parser.c:260: undefined reference to `select_curses_interface'
/home/koeppea/dev/ettercap/src/ec_parser.c:264: undefined reference to `select_gtk_interface'
/home/koeppea/dev/ettercap/src/ec_parser.c:269: undefined reference to `select_daemon_interface'
/home/koeppea/dev/ettercap/src/ec_parser.c:292: undefined reference to `select_text_interface'
libettercap.so.0.0.0: undefined reference to `pcap_findalldevs'
libettercap.so.0.0.0: undefined reference to `pcap_close'
libettercap.so.0.0.0: undefined reference to `pcap_open_live'
libettercap.so.0.0.0: undefined reference to `pcap_geterr'
libettercap.so.0.0.0: undefined reference to `pcap_compile'
libettercap.so.0.0.0: undefined reference to `pcap_stats'
libettercap.so.0.0.0: undefined reference to `pcap_lookupnet'
libettercap.so.0.0.0: undefined reference to `pcap_open_offline'
libettercap.so.0.0.0: undefined reference to `pcap_file'
libettercap.so.0.0.0: undefined reference to `pcap_dump_open'
libettercap.so.0.0.0: undefined reference to `pcap_datalink'
libettercap.so.0.0.0: undefined reference to `pcap_setfilter'
libettercap.so.0.0.0: undefined reference to `pcap_lookupdev'
libettercap.so.0.0.0: undefined reference to `pcap_snapshot'
libettercap.so.0.0.0: undefined reference to `pcap_dump'
libettercap.so.0.0.0: undefined reference to `pcap_fileno'
libettercap.so.0.0.0: undefined reference to `pcap_datalink_val_to_description'
libettercap.so.0.0.0: undefined reference to `pcap_lib_version'
libettercap.so.0.0.0: undefined reference to `pcap_loop'
libettercap.so.0.0.0: undefined reference to `pcap_dump_close'
collect2: error: ld returned 1 exit status
src/CMakeFiles/ettercap.dir/build.make:118: recipe for target 'src/ettercap' failed
make[2]: *** [src/ettercap] Error 1
CMakeFiles/Makefile2:234: recipe for target 'src/CMakeFiles/ettercap.dir/all' failed
make[1]: *** [src/CMakeFiles/ettercap.dir/all] Error 2
Makefile:117: recipe for target 'all' failed
make: *** [all] Error 2
```

File: src/protocols/ec_udp.c [hunk 1]
```diff
@@ -111,7 +111,7 @@ FUNC_DECODER(decode_udp)
           *
           * if the source is the ettercap host, don't display the message 
           */
-         if (ip_addr_is_ours(&PACKET->L3.src) == E_FOUND || ip_addr_is_ours(&PACKET->L3.src) == E_BRIDGE)
+         if (ip_addr_is_ours(&PACKET->L3.src) != E_FOUND && ip_addr_is_ours(&PACKET->L3.src) != E_BRIDGE)
             return NULL;
```
Comment (LocutusOfBorg) [inline src/protocols/ec_udp.c]: @koeppea do you have any clue about this commit I did?
https://github.com/LocutusOfBorg/ettercap/commit/79ceb97c3fee6d2e5924902507a355d4a7597d68
I don't remember the rationale for adding the second comparison to the if...

**Comment by LocutusOfBorg** [comment]
about the interfaces link failure I should have fixed it now
https://github.com/Ettercap/ettercap/pull/698/commits/2da6c13eb40c53711caa2f02bd7ec520d6d25c8a

File: src/protocols/ec_udp.c (hunk 1 above)
Comment (koeppea) [inline src/protocols/ec_udp.c]: Mhhh nope.
Given the purpose of the check described in the comment, the proposed change in this commit does the oposite.
Hence make no sense.

**Comment by koeppea** [comment]
Retested. Now the build succeeds on Linux.
`ettercap` looks good using all interfaces.
`etterlog` and `etterfilter` also behave normal. The only thing with them is that they break my **xterm**:
```
koeppea@lappy:~/tmp$ etterfilter -o testfilter.ef testfilter.ecf

ettercap 0.8.2 copyright 2001-2015 Ettercap Development Team


 14 protocol tables loaded:
	DECODED DATA udp tcp esp gre icmp ipv6 ip arp wifi fddi tr eth 

 13 constants loaded:
	VRRP OSPF GRE UDP TCP ESP ICMP6 ICMP PPTP PPPOE IP6 IP ARP 

 Parsing source file 'testfilter.ecf'  done.

 Unfolding the meta-tree  done.

 Converting labels to real offsets  done.

 Writing output to 'testfilter.ef'  done.

 -> Script encoded into 23 instructions.

koeppea@lappy:~/tmp$ koeppea@lappy:~/tmp$ koeppea@lappy:~/tmp$ koeppea@lappy:~/tmp$ 
```
When executing `etterlog` or `etterfilter` the newline is missing. This usually happens through weird escape sequences. Maybe due to binary output wrongly interpreted.

**Comment by koeppea** [comment]
@LocutusOfBorg 
I've reproduced the build failure of Travis on my MacBook.
```
[ 64%] Building C object src/CMakeFiles/lib_ettercap.dir/interfaces/gtk/ec_gtk_view_profiles.c.o
[ 64%] Linking C shared library libettercap.dylib
Undefined symbols for architecture x86_64:
  "_ec_gbls", referenced from:
      _capture in ec_capture.c.o
      _capture_getifs in ec_capture.c.o
      _load_conf in ec_conf.c.o
      _init_structures in ec_conf.c.o
      _sanity_checks in ec_conf.c.o
      _conntrack_timeouter in ec_conntrack.c.o
      _conntrack_add in ec_conntrack.c.o
      ...
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [src/libettercap.0.0.0.dylib] Error 1
make[1]: *** [src/CMakeFiles/lib_ettercap.dir/all] Error 2
make: *** [all] Error 2
```
Unfortunately, I'm much less experienced with building shared libraries and Cmake.
We could make an appointment and I can give you SSH access to my MacBook so that you could troubleshoot this issue directly on the platform.
What do you think?

**Comment by koeppea** [comment]
Don't know why, but although you merged master into your branch, I'm still missing ff0994b which sill throws the warning that ff0994b fixes.

**Comment by koeppea** [comment]
BTW: Why have you added the last "revert" commit? It looked very promising and very clean.
OSX fails this or the other way. We just have to figure out why.

**Comment by LocutusOfBorg** [comment]
I don't know why I reverted it, not sure if merging the graphical stuff and the library is worth the effort, because people might use ettercap tools (filter, log), with a library built with different graphical support (e.g. from a different computer). this might lead to segfaults, e.g.
try to capture a packet with GEOIP set to true, and open it with etterlog built with no GEOIP support
(I remember having some sort of similar crashes)
for ssh thanks for the offer, I'll try to arrange an evening-debug-session, but I prefer to have some clues before digging into the issue (so far I failed with clang on Ubuntu, I hoped this was a "clang issue", but it seems not)

**Comment by LocutusOfBorg** [comment]
the merge should have added  ff0994b to the branch!

**Comment by koeppea** [comment]
Shall I check again on the Mac?

**Comment by LocutusOfBorg** [comment]
unfortunately I still don't know how to solve...are you on irc?

**Comment by koeppea** [comment]
Just sent you a mail for availability.

**Comment by koeppea** [comment]
I went through the changes.
I think I can continue. However one thing I don't like are the merges from upstream back into your branch.
I'd like to proceed the following:
 1. fire up a new pull request / branch
 2. cherry-pick your commits and apply them into the new branch 
   a. without the merge commits
   b. postpone the travis change
 3. get it working (I think I spotted one or two things)
 4. rebase over master (+resolving conflicts) when outstanding pull requests are merged over master so that merge commits are avoided
 5. apply travis change and adjust if necessary to make travis happy again

Thoughts?

**Comment by sgeto** [comment]
I think if @LocutusOfBorg kept the box that said "Allow edits from maintainers" checked (it's checked by default) when he opened this PR, then you could just take the entire PR over and do a "soft reset".

**Comment by koeppea** [comment]
Yes I could mess with @LocutusOfBorg 's branch, but I itentionally like to avoid that.
One reason is, having a backup or fallback.
The task is not to redo all the work, just to put the bricks together on top of a new foundation.
In case the new approach doesn't work out, @LocutusOfBorg 's work / branch is still untouched.

**Comment by koeppea** [comment]
Closing in favor of reworked pull-request #880

**Pull request closed without merging**